clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -260,12 +262,14 @@ public void handshake() throws IOException {
             maybeThrowSslAuthenticationException();
 
             // this exception could be due to a write. If there is data available to unwrap,
-            // process the data so that any SSLExceptions are reported
+            // process the data so that any SSL handshake exceptions are reported
             if (handshakeStatus == HandshakeStatus.NEED_UNWRAP && netReadBuffer.position() > 0) {
                 try {
                     handshakeUnwrap(false);
-                } catch (SSLException e1) {
+                } catch (SSLHandshakeException | SSLProtocolException | SSLPeerUnverifiedException | SSLKeyException e1) {
                     handshakeFailure(e1, false);
+                } catch (SSLException e1) {
```

Comment:
Shouldn't we be rethrowing the `IOException` in this case as per the comment "If we get here, this is not a handshake failure, throw the original IOException"?

Reply:
Also, I assume the `IOException` we get doesn't provide any clues?

Reply:
No, `IOException` didn't give any clues here. We were falling through to propagate the original exception earlier, but I have added a method to process the exception now.